channelserver/src/server.rs

```diff
@@ -0,0 +1,257 @@
+//! `ChannelServer` is an actor. It maintains list of connection client session.
+//! And manages available channels. Peers send messages to other peers in same
+//! channel through `ChannelServer`.
+
+// use std::sync::{Arc, Mutex};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::time::Instant;
+
+use actix::prelude::{Actor, Context, Handler, Recipient};
+use rand::{self, Rng, ThreadRng};
+use uuid::Uuid;
+
+use logging::MozLogger;
+use perror;
+use settings::Settings;
+
+//pub type ChannelCollection = Arc<Mutex<HashMap<Uuid, HashMap<usize, Channel>>>>;
+//pub type SessionCollection = Arc<Mutex<HashMap<usize,Recipient<TextMessage>>>>;
+
+/// Chat server sends this messages to session
+#[derive(Message)]
+pub struct TextMessage(pub String);
+
+/// Message for chat server communications
+
+/// New chat session is created
+#[derive(Message)]
+#[rtype(usize)]
+pub struct Connect {
+    pub addr: Recipient<TextMessage>,
+    pub channel: Uuid,
+}
+
+/// Session is disconnected
+#[derive(Message)]
+pub struct Disconnect {
+    pub channel: Uuid,
+    pub id: usize,
+}
+
+/// Send message to specific channel
+#[derive(Message)]
+pub struct ClientMessage {
+    /// Id of the client session
+    pub id: usize,
+    /// Peer message
+    pub msg: String,
+    /// channel name
+    pub channel: Uuid,
+}
+
+#[derive(Hash, Eq, PartialEq, Clone, Debug)]
+pub struct Channel {
+    pub id: usize,
+    pub started: Instant,
+    pub msg_count: u8,
+    pub data_exchanged: usize,
+}
+
+/// `ChannelServer` manages chat channels and responsible for coordinating chat
+/// session. implementation is super primitive
+pub struct ChannelServer {
+    me: usize, // id (to see if threading an issue)
+    // collections of sessions grouped by channel
+    channels: HashMap<Uuid, HashMap<usize, Channel>>,
+    // individual connections
+    sessions: HashMap<usize, Recipient<TextMessage>>,
```

Comment:
What's the `usize` used for here, are they all randomly-generated identifiers?  Some sort of typedef may make that easier to follow.

Reply:
I agree that using a type will make this easier to follow. The `usize` is because there are a limited number of sessions that can join a channel.